src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -220,6 +222,30 @@ public static boolean isCorrectToolForDrops(@Nonnull BlockState state, @Nonnull
         return player.hasCorrectToolForDrops(state);
     }
 
+    public static Pair<Predicate<UseOnContext>, Consumer<UseOnContext>> getHoeTillable(UseOnContext context, Pair<Predicate<UseOnContext>, Consumer<UseOnContext>> pair)
+    {
+        Level level = context.getLevel();
+        BlockPos pos = context.getClickedPos();
+        BlockState state = level.getBlockState(pos);
+        // IForgeBlockState#getToolModifiedState has two meanings for null.
+        // One means the event was canceled and another means IForgeBlock's corresponding method returned null.
+        // Since we have a pair variable to default back to, we need to know whether the event was canceled or not.
+        // To do this, we directly call ForgeEventFactory#onToolUse and IForgeBlock#getToolModifiedState.
+        // This is so we can make sure to return null if the event was canceled.
+        BlockState eventState = ForgeEventFactory.onToolUse(state, level, pos, context.getPlayer(), context.getItemInHand(), ToolActions.HOE_DIG);
+        if (eventState == null) // It was canceled
+            return null;
+        BlockState modified;
+        // If the eventState is equal to state, then the event did not change the state, and we should get the tool modified state from the block
+        if (eventState == state) {
+            modified = state.getBlock().getToolModifiedState(state, level, pos, context.getPlayer(), context.getItemInHand(), ToolActions.HOE_DIG);
```

Comment:
This should be its own HOE_TILL ToolAction, not HOE_DIG.
Tool right-click uses are different from their primary "left-click" dig use.

Reply:
Definitely: https://github.com/MinecraftForge/MinecraftForge/blob/bec9d6f53423af875c6f723f261065cffab8c81c/src/main/java/net/minecraftforge/common/ToolActions.java#L84-L87